flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/RootExceptionHistoryEntry.java

```diff
@@ -36,57 +36,91 @@
 
     private static final long serialVersionUID = -7647332765867297434L;
 
-    private final Collection<ExceptionHistoryEntry> concurrentExceptions = new ArrayList<>();
+    private final Iterable<ExceptionHistoryEntry> concurrentExceptions;
+
+    /**
+     * Creates a {@code RootExceptionHistoryEntry} based on the passed {@link
+     * FailureHandlingResultSnapshot}.
+     *
+     * @param snapshot The reason for the failure.
+     * @return The {@code RootExceptionHistoryEntry} instance.
+     * @throws NullPointerException if {@code cause} or {@code failingTaskName} are {@code null}.
+     * @throws IllegalArgumentException if the {@code timestamp} of the passed {@code
+     *     FailureHandlingResult} is not bigger than {@code 0}.
+     */
+    public static RootExceptionHistoryEntry fromFailureHandlingResultSnapshot(
+            FailureHandlingResultSnapshot snapshot) {
+        String failingTaskName = null;
+        TaskManagerLocation taskManagerLocation = null;
+        if (snapshot.getRootCauseExecution().isPresent()) {
+            final Execution rootCauseExecution = snapshot.getRootCauseExecution().get();
+            failingTaskName = rootCauseExecution.getVertexWithAttempt();
+            taskManagerLocation = rootCauseExecution.getAssignedResourceLocation();
+        }
+
+        return new RootExceptionHistoryEntry(
+                snapshot.getRootCause(),
+                snapshot.getTimestamp(),
+                failingTaskName,
+                taskManagerLocation,
+                StreamSupport.stream(snapshot.getConcurrentlyFailedExecution().spliterator(), false)
+                        .map(
+                                execution ->
+                                        ExceptionHistoryEntry.create(
+                                                execution, execution.getVertexWithAttempt()))
+                        .collect(Collectors.toSet()));
```

Comment:
Instead of doing this, could we call `fromGlobalFailure` or `internalCreationMethodSomething` because this part looks quite similar to what the `fromGlobalFailure` does. That way we avoid a bit of code duplication.